uxarray/grid/neighbors.py

```diff
@@ -0,0 +1,220 @@
+            warnings.warn(
+                "Internal representation of face center coordinates has not been tested or verified. Results"
+                "may not be as expected until this has been patched in the Grid class."
+            )
+
+            # center nodes must be present, remove once construction methods for them is added
+            if grid.Mesh2_face_x is None or grid.Mesh2_face_y is None:
+                raise ValueError
+
+            XY = np.vstack((deg2rad(grid.Mesh2_face_y.values),
+                            deg2rad(grid.Mesh2_face_x.values))).T
+
+            self.n_elements = grid.nMesh2_face
+            self.tree = SKBallTree(XY, metric='haversine')
+
+    def query(self,
+              xy: Union[np.ndarray, list, tuple],
+              k: Optional[int] = 1,
+              use_radians: Optional[bool] = False,
+              return_distance: Optional[bool] = True,
+              dualtree: Optional[bool] = False,
+              breadth_first: Optional[bool] = False,
+              sort_results: Optional[bool] = True):
+        """Query the tree for the ``k`` nearest neighbors.
+
+        Parameters
+        ----------
+        xy : array_like
+            coordinate pairs in degrees (lon, lat) to query
+        k: int, default=1
+            The number of nearest neighbors to return
+        use_radians : bool, optional
+            if True, queries assuming xy are inputted in radians, not degrees
+        return_distance : bool, optional
+            if True, return a tuple ``(d, i)`` of distances and indices if False, return array i
+        dualtree : bool, default=False
+            if True, use the dual tree formalism for the query: a tree is built for the query points, and the pair of trees is used to efficiently search this space. This can lead to better performance as the number of points grows large.
+        breadth_first : bool, default=False
+            if True, then query the nodes in a breadth-first manner. Otherwise, query the nodes in a depth-first manner.
+        sort_results : bool, default=True
+            if True, then distances and indices of each point are sorted on return, so that the first column contains the closest points. Otherwise, neighbors are returned in an arbitrary order.
+
+        Returns
+        -------
+        d : ndarray of shape (xy.shape[0], k), dtype=double
+            Each entry gives the list of distances to the neighbors of the corresponding point.
+        ind : ndarray of shape (xy.shape[0], k), dtype=INT_DTYPE
+            Each entry gives the list of indices of neighbors of the corresponding point.
+        """
+
+        if k < 1 or k > self.n_elements:
+            raise AssertionError(
+                f"The value of k must be greater than 1 and less than the number of elements used to construct "
+                f"the tree ({self.n_elemetns}).")
+
+        xy = _prepare_xy_for_query(xy, use_radians)
+
+        # perform query with distance
```

Comment:
Just a question on how this distance works. Is it the arc distance on the sphere from the neighbors? Would this also work on a flat mesh that isn't spherical?

Reply:
The distance metric is computed using the [haversine distance formula](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.haversine_distances.html), which is the great circle distance. 

> Would this also work on a flat mesh that isn't spherical?

I wrote this with the assumption that our grids will represent the surface of a sphere, hence the use of the haversine distance formula. If we have an unstructured grid that isn't on the surface of a sphere, we would need to use a different distance metric.

@erogluorhan @hongyuchen1030 @paullric 

For UXarray, is it fair to assume that we will always be working with grids that represent the surface of a sphere?

Reply:
I tend to say yes since UXarray is dedicated to geoscience data. An alternative could be to provide an optional argument that'd default to haversine distance though.

Reply:
> An alternative could be to provide an optional argument that'd default to haversine distance though.

That's a good idea